test/pool-cl/helpers/CLCustomCurveHook.sol

```diff
@@ -0,0 +1,230 @@
+            Permissions({
+                afterInitialize: true,
+                beforeAddLiquidity: true,
+                beforeSwap: true,
+                befreSwapReturnsDelta: true,
+                // disable all other hooks
+                beforeInitialize: false,
+                afterAddLiquidity: false,
+                beforeRemoveLiquidity: false,
+                afterRemoveLiquidity: false,
+                afterSwap: false,
+                beforeDonate: false,
+                afterDonate: false,
+                afterSwapReturnsDelta: false,
+                afterAddLiquidityReturnsDelta: false,
+                afterRemoveLiquidityReturnsDelta: false
+            })
+        );
+    }
+
+    function afterInitialize(address sender, PoolKey calldata key, uint160 sqrtPriceX96, int24 tick)
+        external
+        override
+        returns (bytes4)
+    {
+        require(sqrtPriceX96 == Constants.SQRT_RATIO_1_1, "InvalidSqrtPrice");
+        require(tick == 0, "InvalidTick");
+        require(key.parameters.getTickSpacing() == 10, "InvalidTickSpacing");
+
+        vault.lock(abi.encode("init", abi.encode(key)));
+
+        return ICLHooks.afterInitialize.selector;
+    }
+
+    function beforeAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        revert("banned adding liquidity by non-hook");
+    }
+
+    function beforeDonate(
+        address sender,
+        PoolKey calldata key,
+        uint256 amount0,
+        uint256 amount1,
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        revert("banned donating by non-hook");
+    }
+
+    function beforeSwap(
+        address,
+        PoolKey calldata key,
+        ICLPoolManager.SwapParams calldata params,
+        bytes calldata hookData
+    ) external override returns (bytes4, BeforeSwapDelta, uint24) {
+        uint256 swapRatio = abi.decode(hookData, (uint256));
+
+        // to simplify the test, amountSpecified is always negative hence always indicates input token amount
+        int256 amountUnSpecified = -params.amountSpecified * int256(swapRatio) / 1000000;
+        BeforeSwapDelta beforeSwapDelta =
+            toBeforeSwapDelta((-params.amountSpecified).toInt128(), -amountUnSpecified.toInt128());
+
+        console2.log("zeroForOne: ", params.zeroForOne);
+        console2.log("amoutSpecified: ", beforeSwapDelta.getSpecifiedDelta());
+        console2.log("amoutUnSpecified: ", beforeSwapDelta.getUnspecifiedDelta());
+
+        if (params.zeroForOne) {
+            // input token is token0, amt = amountSpecified, deposit it to vault
+            key.currency0.take(vault, address(this), (-params.amountSpecified).toUint256(), false);
+            _dispatch(abi.encode("deposit", abi.encode(key, (-params.amountSpecified).toUint256(), 0)));
```

Comment:
Those two could be further combined i.e take() and settle() are exactly opposite direction hence could cancel out which means we could call donate without any settlement here :
1. better gas
2. reserveOfApp is updating but we don't need to have the token in the pool ahead of time